### PR TITLE
fix(builder): cut the spacing bands to a rounded block's curve

### PR DESCRIPTION
**Pull request by mobeenabdullah**
Closes the two rounded-geometry findings left open on the merged #1145
(`PRRT_kwDOSYwUJs6bga9o`, `PRRT_kwDOSYwUJs6bga9q`).

#1154 merged while this was being built, so the base is now `main` and the
branch was rebased onto it — the three commits #1154 landed as a squash are
gone from this branch and only the rounded-geometry commit remains.

## The defect, both directions

A block with `border-radius` curves away from the rectangle its bands are cut
from, and the overlay treated both sides of that as rectangles.

**Outward.** The padding bands are full-width strips taken from the padding
box's rectangle, so a heavily rounded block got colour painted into the
transparent corners outside its rendered border. A band is read as a
MEASUREMENT, so one covering ground the block does not occupy states a
measurement that is false.

**Inward.** A container with `border-radius` and `overflow: hidden` clips on the
curve, so a child inside all four of its padding edges can still lose a corner.
The rectangular comparison accepted it and the bands painted across the part
that is not rendered.

## What was measured rather than assumed

Reaching the USED radii from the computed value is three steps that do not
commute. Each was probed in Chromium:

| Question | Measured |
| --- | --- |
| Do percentages resolve in the computed value? | No — `border-radius: 50%` computes as `"50%"` |
| Is the overlap reduction applied to the computed value? | No — `200px` on a 200×100 box still computes as `200px`, and renders at **50** |
| Is the reduction one factor or one per corner? | **One factor.** `border-top-left-radius: 200px` alone on a 200×100 box draws that corner at 100 — the box's height constraining it through the shared factor |
| Is the padding-box curve reduced a second time when it no longer fits? | **No.** A 200×100 box with a 100px corner and a 10px border draws an inner curve of 90 inside a padding box only 80 tall |
| Does `overflow: clip visible` round its corner? | **No.** The corner is painted in full, where `overflow: hidden` cuts it away |

The last one is why the corner test is asked only when BOTH axes clip: a corner
needs two bounds to exist, and applying the arc on one axis would refuse a block
nothing removes.

Reducing after insetting gives different numbers, so the order is pinned by a
test asserting both the value and that it is not the other one.

## Two decisions worth disagreeing with

**The corner is tested exactly rather than declining every rounded clipping
ancestor.** `border-radius` with `overflow: hidden` is the ordinary card, so a
blanket refusal would blank the overlay for most blocks on a real page. The e2e
test's CONTROL phase is what pins this — a blanket decline fails there, not at
the refusal.

**The fill moved to a pseudo-element.** The band carries the value chip and the
chip deliberately overflows the band it names, so `clip-path` on the band itself
would take the number with it.

## Evidence

Certified by PIXELS, not by attributes. `elementFromPoint` reports a child in
the cut-away corner of an `overflow: hidden` ancestor as hit — measured — so
hit-testing answers about the box tree while the question is what was painted.

Selecting a rounded block must not change a pixel outside it, and the same test
asserts the band DID paint in the middle of the strip, so it cannot pass against
an overlay that drew nothing. With the clip removed that corner pixel reads
`204,235,215` instead of white.

Every new test was break-verified. Two came back GREEN and both were real gaps,
now closed:

- dropping one side from the reduction changed nothing, because every fixture
  let the opposite side bind with the same numbers — there is now one case per
  side, each built so only that side is tight;
- the slack fixture used square corners, so the arc test returned before the
  slack was ever consulted — it now uses a curved corner and asserts the same
  box IS refused with no allowance.

## Gates

Re-run in full after the rebase onto `main`: `pnpm build` 0 · `check-types` 0 ·
builder **1574** · plugin-page-builder 214 · blocks-react 764 · four lints 0 ·
comment convention 0 · changeset 0 ·
`fallow audit` **pass, zero introduced** · e2e **40 passed**.

Two warn-tier styling notes, neither gating. `css-selector-complexity` at the
negative-margin hatch is mine: `[data-negative]` already implies a margin band,
so dropping `[data-box="margin"]` would satisfy the linter, but it would make the
CSS depend on a component invariant instead of saying which band it targets —
left explicit deliberately, and happy to be overruled. The `unused-layer` note is
at line 1 of a file whose first modified line is 1078.

## Summary by CodeRabbit

* **New Features**
  * Spacing overlays now respect rounded corners when displaying margin and padding bands.
  * Rounded clipping preserves value chips and supports elliptical radii, borders, scaling, and nested clipping scenarios.
  * Added accurate handling for blocks partially clipped by rounded ancestor corners.

* **Bug Fixes**
  * Prevented spacing-band fills from painting outside rounded block boundaries.
  * Improved clipping behavior for axis-specific overflow and rounded ancestor edges.

* **Tests**
  * Added comprehensive unit and end-to-end coverage for rounded geometry and overlay rendering.

### CODE REVIEW

**Comment by mobeenabdullah** [comment]
@codex review

**Comment by mobeenabdullah** [comment]
@nextly-bot review

**Comment by mobeenabdullah** [comment]
@codex review

**Comment by mobeenabdullah** [comment]
@nextly-bot review

**Comment by mobeenabdullah** [comment]
@codex review

File: packages/builder/src/spacing-bands.ts [hunk 1]
```diff
@@ -70,17 +76,19 @@ export interface SpacingBand {
...
+   *
+   * Absent on the ordinary square block, and absent on every band that lies
+   * wholly outside the block: a positive margin sits beyond the border edge,
+   * where the margin box is a plain rectangle whatever the corners do, so
+   * nothing there is drawn over ground the block does not occupy.
+   *
+   * Present as INSETS rather than a rectangle because the shape is a different
+   * rectangle from the band, and the band is what carries the clip.
+   */
+  readonly clip?: RoundedInset;
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/builder/src/spacing-bands.ts]: Confirmed and fixed in 1b41beb26 — thank you, this was the sharpest of the round.

Verified before changing anything: `sameBands` compared box, side, label, negative and the four rect fields, and `SpacingOverlay` gates `setBands` on it. A radius-only edit moves none of those, so the overlay kept the array it already had and the fill stayed cut to the previous curve — in both directions, since squaring a rounded block is the same gesture in reverse.

I took the shape of your fix rather than the diff verbatim. The comparison is now split into `sameBand` / `sameRect` / `sameClip`, because folding the clip in as one more `&&` pushed the comparison arrow to CRAP 31.6 and `fallow` refused it. Behaviour is identical to what you proposed.

Covered by seven cases in `spacing-bands.test.ts` — clip appearing, origin, size, one corner's horizontal radius, one corner's vertical radius, clip removed, and two identical clips still comparing equal so the check cannot degenerate into always rebuilding. Break-verified: dropping `sameClip(one.clip, other.clip)` fails six of them.

[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/builder/src/spacing-bands.ts]: Confirmed and fixed in 1b41beb26; same finding as CodeRabbit's on this file, and the P1 rating was right.

The comparison now runs through `sameBand` / `sameRect` / `sameClip` rather than one expression — folding the clip in as another `&&` took the arrow to CRAP 31.6 and `fallow` refused it. Seven cases cover it, including the removal direction and a control that two identical clips still compare equal, so the check cannot degenerate into rebuilding on every measure.

File: packages/builder/src/geometry-dom.ts [hunk 2]
```diff
@@ -704,83 +712,198 @@ export function clippedByAncestor(element: Element, root: Element): boolean {
...
+    insetCornerRadii(outer, {
+      top: edgeWidth(style.borderTopWidth),
+      right: edgeWidth(style.borderRightWidth),
+      bottom: edgeWidth(style.borderBottomWidth),
+      left: edgeWidth(style.borderLeftWidth),
+    }),
+    scale
+  );
+  if (isSquare(inner)) return false;
+  return !boundsInsideRounded(box, clip, inner, slack);
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/builder/src/geometry-dom.ts]: Confirmed and fixed in 1b41beb26. You and Codex found the same defect from opposite ends of the file, and it was the most consequential of the round — the ordinary nested rounded card would have lost its overlay entirely.

`clippedByAncestor` now takes the block's own rendered radii and `roundedInsideRounded` compares two arcs rather than a point against one.

One thing that made this harder than it looks, measured rather than assumed: comparing the two arcs only at their ENDPOINTS is not sufficient. A search over the parameter space found cases where both endpoints of the block's corner arc are inside the container's curve while the middle is not — the pinned fixture is a container arc of 130x24 against a block arc of 60x8, where the worst endpoint sits at f=0.935 and the interior reaches f=1.074. So the arc is sampled at 64 steps, and the sampling bound (`r * (1 - cos(pi/256))`, 0.08px at a 1000px radius) is subtracted from the caller's allowance so the approximation can only ever refuse too much rather than too little.

Certified in the browser as well: a block matched to the container it fills keeps its bands, with the test asserting that the two boxes really do coincide, really do carry the same declaration, and that the block's RECTANGULAR corner really is outside the container's arc — which is what the previous implementation refused on. Break-verified: passing square radii in place of the block's own fails exactly that test.

[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/builder/src/geometry-dom.ts]: Confirmed and fixed in 1b41beb26 — same defect Greptile raised on this line, and your circular-child-in-a-circular-parent example is exactly the case.

Details and the browser evidence are in my reply on that thread. The short version: `clippedByAncestor` now takes the block's own rendered radii, and the comparison samples both arcs rather than testing a rectangular corner against one.

Worth recording from building the fixture, because it caught me first: giving a container and its child the same `border-radius` does NOT give them the same curve. CSS reduces the radii to fit each box, so on the seeded 1280x48 section a declared 60px is drawn at 24 while the 1280x24 child inside it is drawn at 12 — and that child genuinely IS cut. My first version of the test asserted the opposite and failed correctly. The fixture now gives both boxes the same size as well as the same declaration, and asserts both in-test.

File: packages/builder/src/border-radii.ts [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+/**
+ * One computed `<length-percentage>` against the length it is a fraction of.
+ *
+ * Anything unparseable answers zero rather than propagating a `NaN`, which would
+ * poison every comparison downstream into silently answering false.
+ */
+function lengthAgainst(token: string, basis: number): number {
+  const parsed = Number.parseFloat(token);
+  if (!Number.isFinite(parsed)) return 0;
+  const length = token.endsWith("%") ? (parsed / 100) * basis : parsed;
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/builder/src/border-radii.ts:151]: Confirmed reachable, and fixed in 1b41beb26. Measured rather than reasoned: `border-radius: calc(10% + 5px)` computes as `"calc(10% + 5px)"` while `calc(10px + 5px)` computes as `"15px"`, so it is specifically a percentage inside a math function that survives — and the catalog accepts a `calc()` length, so an author reaches it from the inspector.

`usedCornerRadii` now answers `undefined` for a value it cannot resolve, and both consumers treat that as a shape they cannot describe: the overlay draws nothing, and `cutByCorner` refuses the block rather than assuming a square container. That is the answer this package already gives for every other geometry it cannot describe.

One distinction I had to add after the first attempt broke every jsdom test, and I think it is right rather than a workaround: an EMPTY computed value is NOT unknown. A computed style never reports a supported longhand as blank, so a blank one means the engine has no such property — and an engine without `border-radius` draws a square corner, which is what zero says. Treating it as unknown would blank the overlay in any renderer that does not implement the property.

Covered both ways: unit tests for the `calc` form, the resolvable `15px` control, a corner declared `min(10px, 5%)`, and the empty case; plus a browser test that asserts Chromium really did leave the value unresolved before asserting the refusal, so a future Chromium that resolved it would fail loudly rather than pass quietly.

File: packages/builder/src/border-radii.ts [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+export function clipPathOf(inset: RoundedInset): string {
+  const px = (value: number): string =>
+    `${String(Math.round(value * 100) / 100)}px`;
+  const corners = CORNERS.map(corner => inset.radii[corner]);
+  const horizontal = corners.map(corner => px(corner.x)).join(" ");
+  const vertical = corners.map(corner => px(corner.y)).join(" ");
+  const edges = [inset.top, inset.right, inset.bottom, inset.left]
+    .map(px)
+    .join(" ");
+  return `inset(${edges} round ${horizontal} / ${vertical})`;
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/builder/src/border-radii.ts]: You are right, and I verified it rather than taking the spec reading: a 180x80 element under `inset(0px 0px 0px 0px round 90px ...)` is drawn cut at 80, not 90. `inset()` resolves its radii the way `border-radius` does, overlap reduction against the inset rectangle included, so the clip and the element disagreed in exactly the case this module documents.

Fixed in 1b41beb26 by dropping `inset()` for `path()`, which states the arcs outright and is not renormalised. Worth noting for anyone reading later: this is not solvable by pre-scaling the declared radii, because the reduction is a clamp rather than an invertible scaling — `inset()` cannot express a radius larger than its own rectangle at all.

An arc with a zero radius is drawn as a straight line by the SVG path grammar, so square corners need no special case and every shape is written the same way. Break-verified: reverting the emitter to an `inset()` fails three tests, including one asserting the 90px radius survives at full size.

File: packages/builder/src/geometry-dom.ts [hunk 5]
```diff
@@ -704,83 +716,221 @@ export function clippedByAncestor(element: Element, root: Element): boolean {
...
+  const clip = clipEdges(outer, style, scale);
+  if (cutByEdges(box, clip, clipsX, clipsY)) return true;
+  return cutByCorner(
+    box,
+    radii,
+    clip,
+    style,
+    scale,
+    clipsX && clipsY,
+    CLIP_SLACK_PX
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/builder/src/geometry-dom.ts:856]: Confirmed and fixed in f20117b68 — and the measurement found it wider than the inline box.

I probed the catalog's whole display set in Chromium rather than reading a spec table: each value given `overflow: hidden` with a child pulled outside it, then the pixels read. NINE computed display values take no clip:

`inline`, `inline list-item`, `table-row`, `table-row-group`, `table-header-group`, `table-footer-group`, `ruby`, `ruby-text`, `contents`

while `table`, `inline-table` and `table-cell` all clip normally — so a rule written as "anything table-ish" would have been wrong in both directions.

Two things worth recording. `contents` fails twice over and was not in your list: the element generates no box, so nothing clips AND `getBoundingClientRect` answers 0x0 — a descendant compared against a zero-sized rectangle is outside it on every edge, which means that ancestor refused every block even before the corner test existed. That is a pre-existing defect your finding surfaced.

And the measurement itself needed care: my first probe forced overflow with an OVERSIZED child, but table boxes size to content and simply grew to fit, so "not clipped" really meant "never overflowed" — six rows were wrong. A negative-margin child overflows whatever width the ancestor settles on.

`ruby-base`, `ruby-base-container` and `ruby-text-container` are deliberately absent from the set: all three compute to plain `block` in Chromium, so an entry would describe a value no element reports, and there is a test asserting that.

Break-verified: cutting the set down to `inline` alone fails eight cases, and skipping the check entirely fails the browser test that puts a `display: contents` ancestor over a selected block.

File: packages/builder/src/border-radii.ts [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+ * approximation spends the allowance rather than exceeding it.
+ */
+export function roundedInsideRounded(
+  box: EdgeBounds,
+  boxRadii: CornerRadii,
+  clip: EdgeBounds,
+  clipRadii: CornerRadii,
+  slack: number
+): boolean {
+  const allowance = slack - samplingBound(boxRadii);
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/builder/src/border-radii.ts]: Confirmed and fixed in f20117b68. You were right that the fixed count was the problem, and I took your first suggestion.

The sample count is now DERIVED from the radius: enough steps that the chord stays within half the caller's allowance, capped at 512. That fixes the case you name and also stops a four-pixel corner paying the rate a four-thousand-pixel one needs.

One place I went the other way, and I want to be explicit since it is the opposite of "prevent the allowance from turning negative". Past the cap it still can, and I have left it. A negative allowance TIGHTENS every depth rather than loosening it, so the comparison degrades in the safe direction on its own: a shape sitting on the curve is refused, while one comfortably inside is still accepted. I first wrote an explicit `if (allowance < 0) return false` guard for it — break-verification showed it changed no behaviour at all, because the arithmetic already refuses the flush case, and it would have thrown away the correct answer for shapes well inside. So it is gone rather than tested, and there is a test asserting both halves of the degradation.

Also worth noting for the record: this rated P3 but the class of defect is the same one that produced the P1 in the previous round — the overlay silently disappearing on geometry nothing actually clips. Two tests now pin it, one at a 40,000px half-axis and one that checks the chord error is spent OUT of the allowance rather than on top of it. That second one was a green break: nothing had been testing it.

**Comment by mobeenabdullah** [comment]
@codex review

**Comment by mobeenabdullah** [comment]
@nextly-bot review
